### PR TITLE
Indirect development in animals

### PR DESCRIPTION
**Pull request by TaranRallings**
# Description

This includes a few changes to do with animals that undergo indirect developmental, proceeding through a metamorphic process where their functional group changes, such as from caterpillar to butterfly. 

Included:
Animal traits for: development type (direct vs indirect), development status (adult vs larval), and the functional group that is produced by the current functional group through either birth or metamorphosis. This also adds a non-reproductive reproductive type for larval functional groups.

The main methods included are `metamorphose` and `metamorphose_community`, creating the change and processing the change over the entire community, respectively. 

Fixes # (issue)

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: virtual_ecosystem/models/animal/animal_communities.py [hunk 1]
```diff
@@ -353,3 +363,58 @@ def inflict_non_predation_mortality_community(self, dt: timedelta64) -> None:
...
+        # remove the larval cohort
+        larval_cohort.is_alive = False
+        self.remove_dead_cohort(larval_cohort)
+
+    def metamorphose_community(self) -> None:
+        """Handle metamorphosis for all applicable cohorts in the community."""
+
+        for cohort in list(self.all_animal_cohorts):
+            print(f"Processing cohort: {cohort.name}, mass: {cohort.mass_current}")
+            print(f"Development type: {cohort.functional_group.development_type}")
```
Comment (dalonsoa) [inline virtual_ecosystem/models/animal/animal_communities.py:412]: Probably these should be log entries and not print statements.

File: virtual_ecosystem/models/animal/animal_communities.py [hunk 2]
```diff
@@ -353,3 +363,58 @@ def inflict_non_predation_mortality_community(self, dt: timedelta64) -> None:
...
+            if (
+                cohort.functional_group.development_type == DevelopmentType.INDIRECT
+                and (cohort.mass_current >= cohort.functional_group.adult_mass)
+            ):
+                print(f"Metamorphosing cohort: {cohort.name}")
+                self.metamorphose(cohort)
+            else:
+                print(f"Skipping cohort: {cohort.name}")
```
Comment (dalonsoa) [inline virtual_ecosystem/models/animal/animal_communities.py:420]: And these two as well.

File: virtual_ecosystem/models/animal/functional_group.py [hunk 3]
```diff
@@ -129,3 +144,24 @@ def import_functional_groups(
...
+
+    Returns:
+        The FunctionalGroup instance with the matching name.
+
+    Raises:
+        ValueError: If no FunctionalGroup with the given name is found.
+    """
+    for fg in functional_groups:
+        if fg.name == name:
+            return fg
```
Comment (dalonsoa) [inline virtual_ecosystem/models/animal/functional_group.py:166]: I guess there will be only one (or none) groups with the given name, but not more than one, right?

**Review by dalonsoa — APPROVED** [review]
Looks good to me. Just a couple of comments on the `print` statements. 

Are all of those `TODO`s here and there also being open as issues, so they can be tackled in the future?

File: tests/models/animals/test_animal_communities.py [hunk 4]
```diff
@@ -521,3 +525,117 @@ def test_inflict_non_predation_mortality_community(
...
+
+        # Verify
+        number_dead = ceil(
+            larval_cohort.individuals * larval_cohort.constants.metamorph_mortality
+        )
+        expected_individuals = larval_cohort.individuals - number_dead
+
+        animal_community_instance.metamorphose(larval_cohort)
+
+        assert larval_cohort.is_alive is False
```
Comment (alexdewar) [inline tests/models/animals/test_animal_communities.py:561]:
```suggestion
        assert not larval_cohort.is_alive
```

File: tests/models/animals/test_animal_communities.py [hunk 5]
```diff
@@ -521,3 +525,117 @@ def test_inflict_non_predation_mortality_community(
...
+                True,  # Caterpillar should still be alive
+                id="Below_mass_threshold",
+            ),
+            pytest.param(
+                1.1,  # Caterpillar mass is above the adult mass threshold
+                0,  # Caterpillar count should decrease
+                1,  # Butterfly count should increase
+                False,  # Caterpillar should no longer be alive
+                id="Above_mass_threshold",
+            ),
```
Comment (alexdewar) [inline tests/models/animals/test_animal_communities.py:590]: Good job on commenting this so thoroughly!

File: tests/models/animals/test_functional_group.py [hunk 6]
```diff
@@ -213,10 +331,42 @@ def test_import_functional_groups(
...
+
+    if raises_exception:
+        with pytest.raises(
+            ValueError, match=f"No FunctionalGroup with name '{name}' found."
+        ):
+            get_functional_group_by_name(functional_groups, name)
+    else:
+        result = get_functional_group_by_name(functional_groups, name)
+        assert result.name == name
+        assert isinstance(result, FunctionalGroup)
```
Comment (alexdewar) [inline tests/models/animals/test_functional_group.py:372]: For the (hypothetical) situation where `result` doesn't have a `name` attribute, I'd swap the order:

```suggestion
        assert isinstance(result, FunctionalGroup)
        assert result.name == name
```

File: virtual_ecosystem/models/animal/animal_communities.py [hunk 7]
```diff
@@ -353,3 +363,58 @@ def inflict_non_predation_mortality_community(self, dt: timedelta64) -> None:
...
+        self.animal_cohorts[adult_cohort.name].append(adult_cohort)
+
+        # remove the larval cohort
+        larval_cohort.is_alive = False
+        self.remove_dead_cohort(larval_cohort)
+
+    def metamorphose_community(self) -> None:
+        """Handle metamorphosis for all applicable cohorts in the community."""
+
+        for cohort in list(self.all_animal_cohorts):
```
Comment (alexdewar) [inline virtual_ecosystem/models/animal/animal_communities.py:410]: Do you need to convert this to a `list`? Would this work:

```suggestion
        for cohort in self.all_animal_cohorts:
```

File: virtual_ecosystem/models/animal/animal_model.py [hunk 8]
```diff
@@ -177,7 +177,11 @@ def from_config(
         )
 
     def setup(self) -> None:
-        """Method to setup the animal model specific data variables."""
+        """Method to setup the animal model specific data variables.
+
+        TODO: rename this as something else because you've used it crazy
```
Comment (alexdewar) [inline virtual_ecosystem/models/animal/animal_model.py:182]: :laughing:

File: virtual_ecosystem/models/animal/functional_group.py [hunk 9]
```diff
@@ -129,3 +144,24 @@ def import_functional_groups(
     ]
 
     return functional_group_list
+
+
+def get_functional_group_by_name(
+    functional_groups: tuple[FunctionalGroup, ...], name: str
```
Comment (alexdewar) [inline virtual_ecosystem/models/animal/functional_group.py:150]: I guess the input doesn't *have* to be a tuple...

```suggestion
    functional_groups: Iterable[FunctionalGroup], name: str
```

File: virtual_ecosystem/models/animal/functional_group.py [hunk 10]
```diff
@@ -129,3 +144,24 @@ def import_functional_groups(
...
+    Returns:
+        The FunctionalGroup instance with the matching name.
+
+    Raises:
+        ValueError: If no FunctionalGroup with the given name is found.
+    """
+    for fg in functional_groups:
+        if fg.name == name:
+            return fg
+    raise ValueError(f"No FunctionalGroup with name '{name}' found.")
```
Comment (alexdewar) [inline virtual_ecosystem/models/animal/functional_group.py:167]: An alternative:

```suggestion
    return next(fg for fg in functional_groups if fg.name == name)
```

You'll get a `StopIteration` error if there are no matches.

Your way is clear and explicit, so totally up to you which way you do it.

**Review by alexdewar — APPROVED** [review]
I've made some tiny suggestions, but this looks good!

File: virtual_ecosystem/models/animal/animal_communities.py (hunk 1 above)
Comment (TaranRallings) [reply virtual_ecosystem/models/animal/animal_communities.py:412]: Ah, these were for debugging and I left them in accidentally!

File: virtual_ecosystem/models/animal/animal_communities.py [hunk 11]
```diff
@@ -353,3 +363,58 @@ def inflict_non_predation_mortality_community(self, dt: timedelta64) -> None:
...
+        TODO: If the time step drops below a month, this needs an intermediary stage.
+
+        Args:
+            larval_cohort: The cohort in its larval stage to be transformed.
+        """
+
+        # inflict a mortality
+        number_dead = ceil(
+            larval_cohort.individuals * larval_cohort.constants.metamorph_mortality
+        )
```
[the remark this answers is not in the document]
Comment (TaranRallings) [reply virtual_ecosystem/models/animal/animal_communities.py:384]: Yes, that's right. For now, this is an abstraction capturing a range of mortalities from basic metamorphosis failure to predation.

File: virtual_ecosystem/models/animal/functional_group.py (hunk 3 above)
Comment (TaranRallings) [reply virtual_ecosystem/models/animal/functional_group.py:166]: Each functional group has a unique name. It's not clear yet what the final format of the naming system will be but I expect to maintain uniqueness.

File: virtual_ecosystem/models/animal/functional_group.py (hunk 10 above)
Comment (dalonsoa) [reply virtual_ecosystem/models/animal/functional_group.py:167]: Unless the `StopIteration` error is captured and re-raised with some meaningful information, your solution, while concise, might result in very obscure errors.

Comment (alexdewar) [reply virtual_ecosystem/models/animal/functional_group.py:167]: True! I'd only do it this way if I was explicitly checking for errors in the caller.

**Pull request merged**